### PR TITLE
Added GetMyAccount API to AuthenticationClient

### PR DESCRIPTION
**Pull request by andrey-kashcheev**
GetMyAccount API will retrieve information about user:
firstname, lastname, email, phone number, etc.
Functional and integration tests added too.

Relates-To: OLPEDGE-2385
Signed-off-by: Andrey Kashcheev <ext-andrey.kashcheev@here.com>

### CODE REVIEW

File: olp-cpp-sdk-authentication/src/AuthenticationClientImpl.cpp [hunk 1]
```diff
@@ -683,6 +684,49 @@ client::CancellationToken AuthenticationClientImpl::Authorize(
...
+    rapidjson::Document document;
+    rapidjson::IStreamWrapper stream(http_result.response);
+    document.ParseStream(stream);
+    if (http_result.status != http::HttpStatusCode::OK) {
+      // HttpResult response can be error message or valid json with it.
+      std::string msg = http_result.response.str();
+      if (!document.HasParseError() && document.HasMember(Constants::MESSAGE)) {
+        msg = document[Constants::MESSAGE].GetString();
+      }
+      return client::ApiError({http_result.status, msg});
```
Comment (meravingen123) [inline olp-cpp-sdk-authentication/src/AuthenticationClientImpl.cpp:715]: Why there is a '{}' here?

File: olp-cpp-sdk-authentication/src/AuthenticationClientImpl.cpp [hunk 2]
```diff
@@ -683,6 +684,49 @@ client::CancellationToken AuthenticationClientImpl::Authorize(
...
+    }
+
+    if (document.HasParseError()) {
+      return client::ApiError({static_cast<int>(http::ErrorCode::UNKNOWN_ERROR),
+                               "Failed to parse response"});
```
Comment (meravingen123) [inline olp-cpp-sdk-authentication/src/AuthenticationClientImpl.cpp:720]: Why there is a '{}' here?

File: olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp [hunk 3]
```diff
@@ -262,6 +266,54 @@ AuthorizeResult GetAuthorizeResult(rapidjson::Document& doc) {
...
+
+  const NameToMethodVector<UserAccountInfoResponse, time_t> time_fields{
+      {Constants::CREATED_TIME, &UserAccountInfoResponse::SetCreatedTime},
+      {Constants::UPDATED_TIME, &UserAccountInfoResponse::SetUpdatedTime}};
+
+  UserAccountInfoResponse response;
+
+  for (const auto& string_field : string_fields) {
+    if (doc.HasMember(string_field.first))
+      (response.*string_field.second)(doc[string_field.first].GetString());
```
Comment (meravingen123) [inline olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp:301]: In case this element is not string rapidjson will crash

File: olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp [hunk 4]
```diff
@@ -262,6 +266,54 @@ AuthorizeResult GetAuthorizeResult(rapidjson::Document& doc) {
...
+  }
+
+  for (const auto& boolean_field : boolean_fields) {
+    if (doc.HasMember(boolean_field.first))
+      (response.*boolean_field.second)(doc[boolean_field.first].GetBool());
```
Comment (meravingen123) [inline olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp:306]: In case where element is not bool rapidjson will possibly crash

File: olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp [hunk 5]
```diff
@@ -262,6 +266,54 @@ AuthorizeResult GetAuthorizeResult(rapidjson::Document& doc) {
...
+  }
+
+  for (const auto& time_field : time_fields) {
+    if (doc.HasMember(time_field.first))
+      (response.*time_field.second)(doc[time_field.first].GetInt64());
```
Comment (meravingen123) [inline olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp:311]: Same as above

File: olp-cpp-sdk-authentication/src/AuthenticationClientUtils.h [hunk 6]
```diff
@@ -19,17 +19,18 @@
 
 #pragma once
 
+#include <rapidjson/document.h>
+
 #include <string>
```
Comment (meravingen123) [inline olp-cpp-sdk-authentication/src/AuthenticationClientUtils.h:24]: Order of includes

File: olp-cpp-sdk-authentication/include/olp/authentication/Types.h [hunk 7]
```diff
@@ -19,14 +19,15 @@
...
-
 #include <olp/authentication/AuthenticationError.h>
 #include <olp/authentication/AuthorizeResult.h>
 #include <olp/authentication/IntrospectAppResult.h>
+#include <olp/authentication/UserAccountInfoResponse.h>
 #include <olp/core/client/ApiError.h>
 #include <olp/core/client/ApiResponse.h>
 
+#include <memory>
+
```
[the remark this answers is not in the document]
Comment (andrey-kashcheev) [reply olp-cpp-sdk-authentication/include/olp/authentication/Types.h:30]: Reorder was done by clang-format

File: olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp (hunk 3 above)
Comment (andrey-kashcheev) [reply olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp:301]: Yes, as every and each call in this file. Do we need to fix all places?

File: olp-cpp-sdk-authentication/include/olp/authentication/Types.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (andrey-kashcheev) [reply olp-cpp-sdk-authentication/include/olp/authentication/Types.h:30]: Tried to directly format this file, through `git clang-format` and the result is the same as you can see here. I'm using `clang-format-6.0`

[the remark this answers is not in the document]
Comment (andrey-kashcheev) [reply olp-cpp-sdk-authentication/include/olp/authentication/Types.h:30]:
```
IncludeCategories:
  - Regex:           '^<.*\.h>'
    Priority:        1
  - Regex:           '^<.*'
    Priority:        2
```

[the remark this answers is not in the document]
Comment (andescu) [reply olp-cpp-sdk-authentication/include/olp/authentication/Types.h:30]: This should apply only within a block. My clang-format does not format it this way.

[the remark this answers is not in the document]
Comment (andrey-kashcheev) [reply olp-cpp-sdk-authentication/include/olp/authentication/Types.h:30]: What version you are using?

File: olp-cpp-sdk-authentication/.clang-format [hunk 8]
```diff
@@ -1 +0,0 @@
-BasedOnStyle : Google
```
Comment (mykhailo-kuchma) [inline olp-cpp-sdk-authentication/.clang-format:1]: Why did you remove this file?

File: olp-cpp-sdk-authentication/include/olp/authentication/AuthenticationClient.h [hunk 9]
```diff
@@ -568,6 +568,24 @@ class AUTHENTICATION_API AuthenticationClient {
...
+   *
+   * @param access_token A valid access token that is associated with the
+   * user.
+   * @param callback The`UserAccountInfoCallback` method that is called when
+   * the user information request is completed.
+   *
+   * @return The `CancellationToken` instance that can be used to cancel
+   * the request.
+   */
+  client::CancellationToken GetMyAccount(std::string access_token,
```
Comment (mykhailo-kuchma) [inline olp-cpp-sdk-authentication/include/olp/authentication/AuthenticationClient.h:586]: We should also mention that the response depends on token:
Scope Value in Token --> Fields allowed in UserInfo
email --> email, emailVerified, recoveryEmail
openid --> userId, state
phone --> phoneNumber, phoneNumberVerified
profile --> realm, facebookId, firstname, lastname, dob, language, countryCode, marketingEnabled, createdTime, updatedTime, expiresAt

File: olp-cpp-sdk-authentication/src/AuthenticationClientImpl.cpp [hunk 10]
```diff
@@ -683,6 +684,49 @@ client::CancellationToken AuthenticationClientImpl::Authorize(
...
+    }
+
+    return GetUserAccountInfoResponse(document);
```
Comment (mykhailo-kuchma) [inline olp-cpp-sdk-authentication/src/AuthenticationClientImpl.cpp:723]: This can go to a different place, like AuthenticationClientUtils

File: olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp [hunk 11]
```diff
@@ -262,6 +266,54 @@ AuthorizeResult GetAuthorizeResult(rapidjson::Document& doc) {
...
+    if (doc.HasMember(boolean_field.first) && doc[boolean_field.first].IsBool())
+      (response.*boolean_field.second)(doc[boolean_field.first].GetBool());
+  }
+
+  for (const auto& time_field : time_fields) {
+    if (doc.HasMember(time_field.first) && doc[time_field.first].IsInt64())
+      (response.*time_field.second)(doc[time_field.first].GetInt64());
+  }
+
+  return response;
```
Comment (mykhailo-kuchma) [inline olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp:314]: Can we print a warning if there is some field that we don't expect?

**Review by mykhailo-kuchma — CHANGES_REQUESTED** [review]

File: olp-cpp-sdk-authentication/.clang-format (hunk 8 above)
Comment (andrey-kashcheev) [reply olp-cpp-sdk-authentication/.clang-format:1]: This file was removed since it affects clang-format for me. We already have top-level one.

File: olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp (hunk 11 above)
Comment (andrey-kashcheev) [reply olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp:314]: Yes, we can. But I want to mention that checks about a type, about unexpected fields, etc, will be present only here. For me it's strange to see one fortified method and a dozen methods with possible crashes on each 'Get*'. I mean that we need to refactor all those unsafe functions

File: olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp [hunk 12]
```diff
@@ -55,16 +59,164 @@ constexpr auto kOauthSignatureMethod = "oauth_signature_method";
...
+                                                 std::is_enum<T>::value>::type>
+    const char* GetCompatibleType();
+
+    template <typename T, typename = typename std::enable_if<
+                              std::is_same<T, long>::value>::type>
+    int64_t GetCompatibleType();
+
+    using Fields = std::vector<std::pair<
+        const char*, std::function<void(TargetType&, const JsonValue&)>>>;
+
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp:189]: Why not a map? A vector or pairs is nothing else than a bad map.

File: tests/functional/olp-cpp-sdk-authentication/AuthenticationClientTest.cpp [hunk 13]
```diff
@@ -674,4 +700,56 @@ TEST_F(AuthenticationClientTest, IntrospectAppInvalidAccessToken) {
 
   EXPECT_FALSE(response.IsSuccessful());
 }
+
+TEST_F(AuthenticationClientTest, GetMyAccount) {
```
Comment (andescu) [inline tests/functional/olp-cpp-sdk-authentication/AuthenticationClientTest.cpp:704]: Please add SCOPED_TRACE("") for individual tests.

File: tests/integration/olp-cpp-sdk-authentication/AuthenticationClientTest.cpp [hunk 14]
```diff
@@ -1509,3 +1510,93 @@ TEST_F(AuthenticationClientTest, Authorize) {
     testing::Mock::VerifyAndClearExpectations(network_.get());
   }
 }
+
+TEST_F(AuthenticationClientTest, GetMyAccount) {
+  using ResponseType = auth::Response<auth::UserAccountInfoResponse>;
```
Comment (andescu) [inline tests/integration/olp-cpp-sdk-authentication/AuthenticationClientTest.cpp:1515]: Seems you are missing the alias which should be in Types.h

File: olp-cpp-sdk-authentication/include/olp/authentication/Types.h [hunk 15]
```diff
@@ -49,5 +50,8 @@ using AuthorizeResponse = Response<AuthorizeResult>;
 
 /// Called when the user decision request is completed.
 using AuthorizeCallback = Callback<AuthorizeResult>;
+
+/// Called when the get user account request is completed.
+using UserAccountInfoCallback = Callback<UserAccountInfoResponse>;
```
[the remark this answers is not in the document]
Comment (andrey-kashcheev) [reply olp-cpp-sdk-authentication/include/olp/authentication/Types.h:55]: I'm scared to call it UserAccountInfoResponseResponse since UserAccountInfoResponse name already taken

[the remark this answers is not in the document]
Comment (andescu) [reply olp-cpp-sdk-authentication/include/olp/authentication/Types.h:55]: Yeah, this is a little tricky. This is why i proposed to move the UserAccountInfoResponse to `model` namespace, same as we have for read/write. Maybe there is still time to do it :)

File: olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp [hunk 16]
```diff
@@ -262,6 +223,41 @@ AuthorizeResult GetAuthorizeResult(rapidjson::Document& doc) {
...
 
+UserAccountInfoResponse GetUserAccountInfoResponse(
+    client::HttpResponse& http_response) {
+  using UserAccountInfo = model::UserAccountInfoResponse;
+
+  rapidjson::Document document;
+  const auto api_error = Parse(http_response, document);
+  if (!api_error.GetMessage().empty()) {
+    return api_error;
+  }
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/AuthenticationClientUtils.cpp:234]: As you can see this is not the proper way of doing it. Seems you are misusing badly the ApiError class. Return from an API should be always ApiResponse which has the `IsSuccessful()` method to check the outcome of the operation. Why dont you use ApiResponse<rapidjson::Document, ApiError> as return value. Seems rapidjson::Document has move constructor/operator and can be easily movable.

**Review by mykhailo-kuchma — APPROVED** [review]

**Review by andescu — APPROVED** [review]
